File: web/src/lib/transferPathCascade.ts
/** Send-time transfer hops, ordered by expected speed. */
export type TransferHop = 'httpPush' | 'httpPull' | 'webrtc' | 's3';

export type TransferChannel = 'lan' | 'webrtc' | 's3';

/** Runtime UI phase while the send cascade probes and then transfers. */
export type TransferPhase =
  | 'tryingHttp'
  | 'waitingPull'
  | 'connectingWebrtc'
  | 'connectingWebrtcFallback'
  | 'tryingS3'
  | 'tryingS3Fallback'
  | 'sendingHttp'
  | 'sendingWebrtc'
  | 'sendingS3';

export function isConnectingTransferPhase(phase?: string | null): boolean {
  return (
    phase === 'tryingHttp' ||
    phase === 'waitingPull' ||
    phase === 'connectingWebrtc' ||
    phase === 'connectingWebrtcFallback' ||
    phase === 'tryingS3' ||
    phase === 'tryingS3Fallback'
  );
}

export function channelOfTransferPhase(phase?: string | null): TransferChannel | undefined {
  switch (phase) {
    case 'tryingHttp':
    case 'waitingPull':
    case 'sendingHttp':
      return 'lan';
    case 'connectingWebrtc':
    case 'connectingWebrtcFallback':
    case 'sendingWebrtc':
      return 'webrtc';
    case 'tryingS3':
    case 'tryingS3Fallback':
    case 'sendingS3':
      return 's3';
    default:
      return undefined;
  }
}

export function fileTransferTypeFromPayload(payload: unknown): TransferChannel | undefined {
  if (!payload || typeof payload !== 'object') return undefined;
  const p = payload as {
    lan?: boolean;
    webrtc?: boolean;
    key?: string;
    targetDeviceIds?: unknown;
  };
  if (p.lan === true) return 'lan';
  if (p.webrtc === true) return 'webrtc';
  if (typeof p.key === 'string' && p.key.length > 0) return 's3';
  if (Array.isArray(p.targetDeviceIds) && p.targetDeviceIds.length > 0) return 'lan';
  return undefined;
}

export type TransferPathInput = {
  localIsWeb: boolean;
  peerIsWeb: boolean;
  isLoggedIn: boolean;
  webrtcAvailable: boolean;
  s3Configured: boolean;
  s3Online: boolean;
  isS3VirtualSession?: boolean;
};

/**
 * Hops that can be attempted for this pair, fastest first.
 *
 * Direction rules:
 * - Web → App: httpPush only (browser cannot serve reverse-pull)
 * - App → Web: httpPull only (browser cannot receive POST /transfer)
 * - Web → Web: skip HTTP
 * - Guest: same HTTP direction; S3 omitted
 */
export function applicableTransferHops(input: TransferPathInput): TransferHop[] {
  if (input.isS3VirtualSession) {
    if (input.isLoggedIn && input.s3Configured && input.s3Online) return ['s3'];
    return [];
  }

  const hops: TransferHop[] = [];
  if (!input.peerIsWeb) hops.push('httpPush');
  if (!input.localIsWeb) hops.push('httpPull');
  if (input.webrtcAvailable) hops.push('webrtc');
  if (input.isLoggedIn && input.s3Configured && input.s3Online) hops.push('s3');
  return hops;
}

export function transferChannelLabel(type?: string | null): string {
  switch (type) {
    case 'lan':
      return 'HTTP';
    case 'webrtc':
      return 'WebRTC';
    case 's3':
      return 'S3';
    default:
      return '';
  }
}

export function transferPhaseMessageKey(phase?: string | null): string | undefined {
  switch (phase) {
    case 'tryingHttp':
      return 'chat.bubble.phaseTryingHttp';
    case 'waitingPull':
      return 'chat.bubble.phaseWaitingPull';
    case 'connectingWebrtc':
      return 'chat.bubble.phaseConnectingWebrtc';
    case 'connectingWebrtcFallback':
      return 'chat.bubble.phaseConnectingWebrtcFallback';
    case 'tryingS3':
      return 'chat.bubble.phaseTryingS3';
    case 'tryingS3Fallback':
      return 'chat.bubble.phaseTryingS3Fallback';
    default:
      return undefined;
  }
}

/** Directed `toDeviceId` wins; otherwise match `payload.targetDeviceIds`. */
export function isLanFileOfferForMe(opts: {
  me: string;
  toDeviceId?: string | null;
  targetDeviceIds?: unknown;
}): boolean {
  const { me, toDeviceId, targetDeviceIds } = opts;
  if (typeof toDeviceId === 'string' && toDeviceId.length > 0) {
    return toDeviceId === me;
  }
  return Array.isArray(targetDeviceIds) && targetDeviceIds.includes(me);
}

export function hopsIncludeHttp(hops: TransferHop[]): boolean {
  return hops.includes('httpPush') || hops.includes('httpPull');
}

export class TransferHopSkipCache {
  private readonly failedAt = new Map<string, Map<TransferHop, number>>();

  constructor(private readonly ttlMs = 20_000) {}

  markFailed(peerId: string, hop: TransferHop): void {
    let byHop = this.failedAt.get(peerId);
    if (!byHop) {
      byHop = new Map();
      this.failedAt.set(peerId, byHop);
    }
    byHop.set(hop, Date.now());
  }

  markSucceeded(peerId: string, hop: TransferHop): void {
    this.failedAt.get(peerId)?.delete(hop);
  }

  shouldSkip(peerId: string, hop: TransferHop): boolean {
    const at = this.failedAt.get(peerId)?.get(hop);
    if (at == null) return false;
    if (Date.now() - at > this.ttlMs) {
      this.failedAt.get(peerId)?.delete(hop);
      return false;
    }
    return true;
  }

  filter(peerId: string, hops: TransferHop[]): TransferHop[] {
    return hops.filter((hop) => !this.shouldSkip(peerId, hop));
  }
}

File: app/lib/network/transfer_path_cascade.dart
/// Send-time transfer hops, ordered by expected speed.
///
/// This is a priori ranking plus connectivity attempts — not a bandwidth race.
enum TransferHop {
  /// Sender POSTs to the receiver's LAN HTTP server.
  httpPush,

  /// Receiver GETs from the sender's LAN HTTP server (`lan_file_offer`).
  httpPull,

  /// WebRTC DataChannel.
  webrtc,

  /// S3-compatible relay. Logged-in only.
  s3,
}

/// Runtime UI phase while the send cascade probes and then transfers.
abstract final class TransferPhase {
  static const tryingHttp = 'tryingHttp';
  static const waitingPull = 'waitingPull';
  static const connectingWebrtc = 'connectingWebrtc';
  static const connectingWebrtcFallback = 'connectingWebrtcFallback';
  static const tryingS3 = 'tryingS3';
  static const tryingS3Fallback = 'tryingS3Fallback';
  static const sendingHttp = 'sendingHttp';
  static const sendingWebrtc = 'sendingWebrtc';
  static const sendingS3 = 'sendingS3';

  static bool isConnecting(String? phase) {
    return phase == tryingHttp ||
        phase == waitingPull ||
        phase == connectingWebrtc ||
        phase == connectingWebrtcFallback ||
        phase == tryingS3 ||
        phase == tryingS3Fallback;
  }

  static String? channelOf(String? phase) {
    return switch (phase) {
      tryingHttp || waitingPull || sendingHttp => 'lan',
      connectingWebrtc ||
      connectingWebrtcFallback ||
      sendingWebrtc => 'webrtc',
      tryingS3 || tryingS3Fallback || sendingS3 => 's3',
      _ => null,
    };
  }
}

class TransferPathInput {
  const TransferPathInput({
    required this.localIsWeb,
    required this.peerIsWeb,
    required this.isLoggedIn,
    required this.webrtcAvailable,
    required this.s3Configured,
    required this.s3Online,
    this.isS3VirtualSession = false,
  });

  /// Browser clients cannot serve LAN HTTP (no reverse-pull, cannot receive push).
  final bool localIsWeb;
  final bool peerIsWeb;
  final bool isLoggedIn;
  final bool webrtcAvailable;
  final bool s3Configured;
  final bool s3Online;
  final bool isS3VirtualSession;
}

/// Whether this device should act on a `lan_file_offer`.
///
/// Directed envelopes (`toDeviceId`) win so guest `device-send` and
/// logged-in 1:1 signaling reach the intended peer. Broadcast offers still
/// match [targetDeviceIds].
bool isLanFileOfferForMe({
  required String me,
  String? toDeviceId,
  Object? targetDeviceIds,
}) {
  if (toDeviceId != null && toDeviceId.isNotEmpty) {
    return toDeviceId == me;
  }
  if (targetDeviceIds is List) {
    return targetDeviceIds.contains(me);
  }
  return false;
}

/// Hops that can be attempted for this pair, fastest first.
///
/// Direction rules:
/// - Web → App: [TransferHop.httpPush] only (no pull)
/// - App → Web: [TransferHop.httpPull] only (no push)
/// - Web → Web: skip HTTP
/// - Guest: same HTTP direction; S3 omitted
List<TransferHop> applicableTransferHops(TransferPathInput input) {
  if (input.isS3VirtualSession) {
    if (input.isLoggedIn && input.s3Configured && input.s3Online) {
      return const [TransferHop.s3];
    }
    return const [];
  }

  final hops = <TransferHop>[];

  // Web has no LAN HTTP server, so it cannot receive POST /transfer.
  if (!input.peerIsWeb) {
    hops.add(TransferHop.httpPush);
  }
  // Web cannot expose files for reverse pull.
  if (!input.localIsWeb) {
    hops.add(TransferHop.httpPull);
  }

  if (input.webrtcAvailable) {
    hops.add(TransferHop.webrtc);
  }

  if (input.isLoggedIn && input.s3Configured && input.s3Online) {
    hops.add(TransferHop.s3);
  }

  return hops;
}

bool hopsIncludeHttp(List<TransferHop> hops) {
  return hops.contains(TransferHop.httpPush) ||
      hops.contains(TransferHop.httpPull);
}

/// Short-TTL skip of hops that just failed so the next send does not stall
/// on a path that is known-down. Unknown / expired → try.
class TransferHopSkipCache {
  TransferHopSkipCache({this.ttl = const Duration(seconds: 20)});

  final Duration ttl;
  final Map<String, Map<TransferHop, DateTime>> _failedAt = {};

  void markFailed(String peerId, TransferHop hop) {
    _failedAt.putIfAbsent(peerId, () => {})[hop] = DateTime.now();
  }

  void markSucceeded(String peerId, TransferHop hop) {
    _failedAt[peerId]?.remove(hop);
  }

  bool shouldSkip(String peerId, TransferHop hop) {
    final at = _failedAt[peerId]?[hop];
    if (at == null) return false;
    if (DateTime.now().difference(at) > ttl) {
      _failedAt[peerId]?.remove(hop);
      return false;
    }
    return true;
  }

  List<TransferHop> filter(String peerId, List<TransferHop> hops) {
    return hops.where((hop) => !shouldSkip(peerId, hop)).toList(growable: false);
  }
}
